### PR TITLE
Update unit tests of kmeans, pca, factorization machines, lda and ntm

### PR DESCRIPTION
**Pull request by yangaws**
Description:
1, Remove unused codes in unit tests of ntm and lda
2, Add more tests in unit test of factorization machines
3, Add pca and kmeans unit tests
4, Fix type of hyper-parameter tol in kmeans
5, Add missing hyper-parameter eval_metrics in kmeans
6, Remove some tests in test_amazon_estimator since they are in unit test of pca
7, Use validator function in pca instead of lambda

Test:
tox tests/unit passed

### CODE REVIEW

File: src/sagemaker/amazon/kmeans.py [hunk 1]
```diff
@@ -27,16 +27,18 @@ class KMeans(AmazonAlgorithmEstimatorBase):
...
     init_method = hp('init_method', isin('random', 'kmeans++'), 'One of "random", "kmeans++"', str)
     max_iterations = hp('local_lloyd_max_iterations', gt(0), 'An integer greater-than 0', int)
-    tol = hp('local_lloyd_tol', gt(0), 'An integer greater-than 0', int)
+    tol = hp('local_lloyd_tol', (ge(0), le(1)), 'An float in [0, 1]', float)
     num_trials = hp('local_lloyd_num_trials', gt(0), 'An integer greater-than 0', int)
     local_init_method = hp('local_lloyd_init_method', isin('random', 'kmeans++'), 'One of "random", "kmeans++"', str)
     half_life_time_size = hp('half_life_time_size', ge(0), 'An integer greater-than-or-equal-to 0', int)
     epochs = hp('epochs', gt(0), 'An integer greater-than 0', int)
     center_factor = hp('extra_center_factor', gt(0), 'An integer greater-than 0', int)
+    eval_metrics = hp(name='eval_metrics', validation_message='A comma separated list of "msd" or "ssd"',
```
Comment (winstonaws) [inline src/sagemaker/amazon/kmeans.py]: Where did you get comma separated list from? The API docs seem to imply just one value: https://docs.aws.amazon.com/sagemaker/latest/dg/k-means-api-config.html

If the API docs are wrong, can you ask the algorithms team to fix the docs?

File: src/sagemaker/amazon/pca.py [hunk 2]
```diff
@@ -25,13 +26,11 @@ class PCA(AmazonAlgorithmEstimatorBase):
 
     DEFAULT_MINI_BATCH_SIZE = 500
 
-    num_components = hp(name='num_components', validate=lambda x: x > 0,
-                        validation_message='Value must be an integer greater than zero', data_type=int)
-    algorithm_mode = hp(name='algorithm_mode', validate=lambda x: x in ['regular', 'stable', 'randomized'],
-                        validation_message='Value must be one of "regular", "stable", "randomized"', data_type=str)
+    num_components = hp('num_components', gt(0), 'Value must be an integer greater than zero', int)
+    algorithm_mode = hp('algorithm_mode', isin('regular', 'stable', 'randomized'),
```
Comment (winstonaws) [inline src/sagemaker/amazon/pca.py]: Same here: https://docs.aws.amazon.com/sagemaker/latest/dg/PCA-reference.html

File: src/sagemaker/amazon/pca.py [hunk 3]
```diff
@@ -25,13 +26,11 @@ class PCA(AmazonAlgorithmEstimatorBase):
...
+                        'Value must be one of "regular", "stable", "randomized"', str)
     subtract_mean = hp(name='subtract_mean', validation_message='Value must be a boolean', data_type=bool)
-    extra_components = hp(name='extra_components', validate=lambda x: x >= 0,
-                          validation_message="Value must be an integer greater than or equal to 0", data_type=int)
+    extra_components = hp('extra_components', ge(0), "Value must be an integer greater than or equal to 0", int)
```
Comment (winstonaws) [inline src/sagemaker/amazon/pca.py]: Maybe edit the description to state that you should leave this unset if you want the behavior for -1: https://docs.aws.amazon.com/sagemaker/latest/dg/PCA-reference.html

File: tests/unit/test_fm.py [hunk 4]
```diff
@@ -94,3 +108,282 @@ def test_all_hyperparameters(sagemaker_session):
 def test_image(sagemaker_session):
     fm = FactorizationMachines(sagemaker_session=sagemaker_session, **ALL_REQ_ARGS)
     assert fm.train_image() == registry(REGION) + '/factorization-machines:1'
+
+
+def test_num_factors_validation_fail_type(sagemaker_session):
```
Comment (winstonaws) [inline tests/unit/test_fm.py]: I'd recommend using parameterized test functions to make these tests significantly more concise for each: https://docs.pytest.org/en/latest/parametrize.html

We can chat about the best way to do this if you like.

File: tests/unit/test_fm.py [hunk 5]
```diff
@@ -94,3 +108,282 @@ def test_all_hyperparameters(sagemaker_session):
...
+
+    fm.fit(data, MINI_BATCH_SIZE)
+
+    base_fit.assert_called_once()
+    assert len(base_fit.call_args[0]) == 2
+    assert base_fit.call_args[0][0] == data
+    assert base_fit.call_args[0][1] == MINI_BATCH_SIZE
+
+
+def test_call_fit_none_mini_batch_size(sagemaker_session):
```
Comment (winstonaws) [inline tests/unit/test_fm.py]: What is this test asserting on? And does the base fit need to be patched? (Same question about patch applies to the tests below as well)

File: tests/unit/test_fm.py (hunk 4 above)
Comment (yangaws) [reply tests/unit/test_fm.py]: As discussed offline, all hyper-parameter related unit tests are now in parametrized style.

**Pull request merged**